### PR TITLE
Switch to meson subprojects/wraps

### PR DESCRIPTION
**Pull request by ret2libc**
**DO NOT MERGE YET, see https://github.com/rizinorg/rizin/pull/510**

 **Detailed description**

This PR makes use of meson subprojects. It removes git submodules.

Advantages:
- no need to mess with git submodules, --recurse-submodules, etc.
- easy to download external deps, because you don't need to check if tar/wget/curl are available (in Windows they are probably not), so it can work easily on multiple platforms
- reduce the complexity of the meson.build files, because they are more focused on a single subproject
- there are already wraps available in Meson WrapDB for lz4, zlib, libzip, xxhash. Other stuff could be moved to subprojects as well easily.

Possible disadvantages:
- once a subproject is downloaded, it is not automatically updated if you update the .wrap file. You have to either delete the old `subprojects/X` directory or run `meson subproject update`.

### CODE REVIEW

**Comment by ret2libc** [comment]
This closes https://github.com/rizinorg/rizin/issues/188

**Comment by XVilka** [comment]
It looks much better. I will test it locally, and if it works on my setups - let's merge this. 👍

**Comment by XVilka** [comment]
It doesn't solve the problem of "git clone" - it still uses it unconditionally, thus https://github.com/rizinorg/rizin/issues/188 issue still persists.

**Comment by ret2libc** [comment]
What command are you running? If you run meson with -Duse_capstone_version=v4 i doubt it uses git clone

**Comment by XVilka** [comment]
Yes, this way it works, but there is no way to fetch latest capstone from git repository this way.
What about creating a copy of `subprojects/capstone-bundled.wrap` and `subprojects/capstone-next.wrap` using direct URL like `subprojects/capstone-v4.wrap` does? Call them `subprojects/capstone-bundled-archive.wrap` and `subprojects/capstone-next-archive.wrap` maybe? Same with SDB.

**Comment by XVilka** [comment]
The rest is LGTM.

**Comment by XVilka** [comment]
But seems there is a problem on MacOS:
```
[44/1485] Linking target subprojects/sdb/libsdb.1.5.1-git.dylib
FAILED: subprojects/sdb/libsdb.1.5.1-git.dylib 
clang  -o subprojects/sdb/libsdb.1.5.1-git.dylib subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_array.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_set.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_base64.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_buffer.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_cdb.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_cdb_make.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_dict.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_diff.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_disk.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_fmt.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_uu.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_pp.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ht_up.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_journal.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_json.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_lock.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ls.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_match.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_ns.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_num.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_query.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_sdb.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_sdbht.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_util.c.o subprojects/sdb/libsdb.1.5.1-git.dylib.p/src_text.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,error -shared -install_name @rpath/libsdb.1.5.1-git.dylib -compatibility_version 1.5.1-git -current_version 1.5.1-git
ld: malformed 32-bit x.y.z version number: 1.5.1-git
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by ret2libc** [comment]
> Yes, this way it works, but there is no way to fetch latest capstone from git repository this way.
> What about creating a copy of `subprojects/capstone-bundled.wrap` and `subprojects/capstone-next.wrap` using direct URL like `subprojects/capstone-v4.wrap` does? Call them `subprojects/capstone-bundled-archive.wrap` and `subprojects/capstone-next-archive.wrap` maybe? Same with SDB.

I understand your problems with the firewall, but IMO we should not go too far and duplicate everything. That means more things to test and we already support a lot of configurations. In your case I would either just use `-Duse_capstone_version=v4` and stick with it or manually download the archive from our repository `rizinorg/capstone` in the `subprojects/` directory.

**Comment by XVilka** [comment]
There are millions of people who will need this. Not just me.

**Comment by ret2libc** [comment]
> There are millions of people who will need this. Not just me.

Then they should stick with releases probably. We are a project based on `git`, not using `git` at all seems ... weird. If you can't use `git`, you can't develop. If you can't develop, you just stick with releases that just involves downloading a tarball and unzipping it.

**Comment by XVilka** [comment]
It's possible to use `git`, it's just cloning `capstone` is slow due to the repository/history size. Anyway, if you don't want to do this - I can send my own PR.

**Comment by ret2libc** [comment]
I will need https://github.com/rizinorg/sdb/pull/5 to continue.

**Comment by XVilka** [comment]
Seems like broken VS build:
```
OLDNAMES.lib(unlink.obi) : error LNK2001: unresolved external symbol __imp_unlink
librz_analysis.a(src_cdb_make.c.obj) : error LNK2019: unresolved external symbol __imp__aligned_free referenced in function cdb_alloc_free
librz_analysis.a(src_cdb_make.c.obj) : error LNK2019: unresolved external symbol __imp__aligned_malloc referenced in function cdb_alloc
librz_analysis.a(src_query.c.obj) : error LNK2019: unresolved external symbol __imp_puts referenced in function sdb_query
OLDNAMES.lib(unlink.obi) : error LNK2001: unresolved external symbol __imp__unlink
test/unit/test_sign.exe : fatal error LNK1120: 5 unresolved externals
ninja: build stopped: subcommand failed.
Command exited with code 1
```

**Review by thestr4ng3r — APPROVED** [review]
lgtm after rebase and ci fix

File: meson.build
Comment (thestr4ng3r) [inline meson.build]: either run this or wat

Comment (ret2libc) [reply meson.build]: ahh :D old msg !

**Comment by XVilka** [comment]
Seems there is a problem:
```
FAILED: librz/cons/rz_cons.dll 
"link"  /MACHINE:x64 /OUT:librz/cons/rz_cons.dll librz/cons/rz_cons.dll.p/canvas.c.obj librz/cons/rz_cons.dll.p/canvas_line.c.obj librz/cons/rz_cons.dll.p/cons.c.obj librz/cons/rz_cons.dll.p/html.c.obj librz/cons/rz_cons.dll.p/editor.c.obj librz/cons/rz_cons.dll.p/grep.c.obj librz/cons/rz_cons.dll.p/hud.c.obj librz/cons/rz_cons.dll.p/input.c.obj librz/cons/rz_cons.dll.p/less.c.obj librz/cons/rz_cons.dll.p/line.c.obj librz/cons/rz_cons.dll.p/more.c.obj librz/cons/rz_cons.dll.p/output.c.obj librz/cons/rz_cons.dll.p/pager.c.obj librz/cons/rz_cons.dll.p/pal.c.obj librz/cons/rz_cons.dll.p/cpipe.c.obj librz/cons/rz_cons.dll.p/rgb.c.obj librz/cons/rz_cons.dll.p/cutf8.c.obj "/nologo" "/release" "/nologo" "/OPT:REF" "/DLL" "/IMPLIB:librz/cons\rz_cons.lib" "librz/util\rz_util.lib" "librzspp.a" "subprojects\sdb\src\libsdb.a" "librzzlib.a" "ws2_32.lib" "wininet.lib" "psapi.lib" "kernel32.lib" "user32.lib" "gdi32.lib" "winspool.lib" "shell32.lib" "ole32.lib" "oleaut32.lib" "uuid.lib" "comdlg32.lib" "advapi32.lib"
libsdb.a(sdb.c.obj) : error LNK2005: sdb_foreach_list already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_new0 already defined in rz_util.lib(rz_util.dll)
libsdb.a(sdb.c.obj) : error LNK2005: sdb_set already defined in rz_util.lib(rz_util.dll)
libsdb.a(ls.c.obj) : error LNK2005: ls_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_find already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_free already defined in rz_util.lib(rz_util.dll)
libsdb.a(ht_pp.c.obj) : error LNK2005: ht_pp_new already defined in rz_util.lib(rz_util.dll)
   Creating library librz/cons\rz_cons.lib and object librz/cons\rz_cons.exp
librz/cons/rz_cons.dll : fatal error LNK1169: one or more multiply defined symbols found
[713/1489] Generating symbol file librz/socket/rz_socket.dll.p/rz_socket.dll.symbols
[714/1489] Generating symbol file librz/crypto/rz_crypto.dll.p/rz_crypto.dll.symbols
ninja: build stopped: subcommand failed.
Command exited with code 1
```
https://ci.appveyor.com/project/rizinorg/rizin/builds/36938607/job/mwcnf204b7b8hknh#L1275

**Comment by XVilka** [comment]
There are many conflicts now, please rebase when you have time.

**Comment by ret2libc** [comment]
> There are many conflicts now, please rebase when you have time.

Yeah I'm waiting for first release, I will revive this after that.

File: doc/PACKAGERS.md
Comment (ret2libc) [inline doc/PACKAGERS.md:26]: add mention of `meson dist`

File: .github/workflows/ci.yml
Comment (ret2libc) [inline .github/workflows/ci.yml:431]: review if it's really necessary `static_runtime`. Maybe just remove in favour of this.

File: meson_options.txt
Comment (ret2libc) [inline meson_options.txt]: maybe use affermative option instead of negative one. `option(sub-projects_check, value: true, ..)`

File: doc/PACKAGERS.md
Comment (ret2libc) [reply doc/PACKAGERS.md:26]: Or maybe not, since you have to have anyway a build directory to run `meson dist`.

File: .github/workflows/clang-analysis.yml
Comment (ghost) [inline .github/workflows/clang-analysis.yml:20]: You can replace line 54
```
        grep -Rl 'rizin/shlr/capstone' reports | xargs rm -fv
        grep -Rl 'rizin/shlr/sdb' reports | xargs rm -fv
        grep -Rl 'rizin/shlr/tree-sitter' reports | xargs rm -fv
```

with

```
        grep -Rl 'rizin/subprojects' reports | xargs rm -fv
```

Comment (ret2libc) [reply .github/workflows/clang-analysis.yml:20]: Thanks, done!

File: shlr/meson.build
Comment (kazarmy) [inline shlr/meson.build:67]: We don't need to patch capstone anymore?

File: doc/capstone.md
Comment (kazarmy) [inline doc/capstone.md:14]: If no patches are applied then the parenthesized phrase can be removed.

File: shlr/meson.build
Comment (ret2libc) [reply shlr/meson.build:67]: See capstone bundled wrap. We have a fork there and i already applied the patches in the fork

File: doc/capstone.md
Comment (ret2libc) [reply doc/capstone.md:14]: i'll keep the parentheses because the patches are still applied, they just are directly in the git repo.

**Review by kazarmy — APPROVED** [review]
lgtm

**Review by thestr4ng3r — APPROVED** [review]

**Review by wargio — APPROVED** [review]

**Review by XVilka — APPROVED** [review]
LGTM. But could you please more CI tasks by pushing into the `dist-` branch maybe? So we can check other build jobs too.

**Comment by ret2libc** [comment]
**DO NOT MERGE YET, see #510**

**Comment by ret2libc** [comment]
All green with `dist`, see https://github.com/rizinorg/rizin/pull/510

**Review by XVilka — APPROVED** [review]
LGTM, lets merge this

**Pull request merged**
